src/Middleware/Rewrite/src/RewriteBuilderExtensions.cs

```diff
@@ -46,7 +48,46 @@ public static IApplicationBuilder UseRewriter(this IApplicationBuilder app, Rewr
             }
 
             // put middleware in pipeline
-            return app.UseMiddleware<RewriteMiddleware>(Options.Create(options));
+            return AddRewriteMiddleware(app, Options.Create(options));
+        }
+
+        private static IApplicationBuilder AddRewriteMiddleware(IApplicationBuilder app, IOptions<RewriteOptions>? options)
+        {
+            const string globalRouteBuilderKey = "__GlobalEndpointRouteBuilder";
+            // Only use this path if there's a global router (in the 'WebApplication' case).
+            if (app.Properties.TryGetValue(globalRouteBuilderKey, out var routeBuilder) && routeBuilder is not null)
+            {
+                return app.Use(next =>
+                {
+                    if (options is null)
+                    {
+                        options = app.ApplicationServices.GetRequiredService<IOptions<RewriteOptions>>();
+                    }
+
+                    var webHostEnv = app.ApplicationServices.GetRequiredService<IWebHostEnvironment>();
+                    var loggerFactory = app.ApplicationServices.GetRequiredService<ILoggerFactory>();
+
+                    // start a new middleware pipeline
+                    var builder = app.New();
+                    builder.UseMiddleware<RewriteMiddleware>(options);
```

Comment:
```suggestion
```
This doesn't look right, why do we need to re-run the rewrite middleware on the new pipeline? Is this leftover?